31 open memory

songmeo opened the pull request.

songmeo committed:
feed the file to conversation model

songmeo committed:
upload file tool call

songmeo committed:
address review comments

songmeo committed:
address comments

songmeo committed:
add openmemory

songmeo committed:
fix mypy, move adding item to context part into process_output

songmeo committed:
remove user_id filter from get_memory

songmeo committed:
add more info logs, add direct channel check:

songmeo committed:
move memory to a separate module

songmeo committed:
address comments

songmeo committed:
add memory tags

songmeo committed:
add tags by sectors and keywords

pavel-kirienko committed:
fix dependency specs

pavel-kirienko commented:
Update the README

pavel-kirienko commented:
Update the slack token generation instructions

pavel-kirienko commented:
Document environment variables, incl. BRO_USER_ID

pavel-kirienko commented:
Non-standard env vars need to have BRO prefix; the user-ID needs to mention that it is specific to Slack. The correct names are:

- `BRO_SLACK_BOT_TOKEN`
- `BRO_SLACK_APP_TOKEN`
- `BRO_SLACK_USER_ID`

pavel-kirienko commented:
**First issue** -- the conversation layer doesn't understand that interaction needs to be carried on:

>03:20:35 WAR bro.conversation: 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 🏁 
>I will now select a cat image without visible watermarks, open it, download it, then switch to Slack and send it to the specified user/channel with a short friendly message.

The conversation should have said "okay go on" but it just stopped.

**Second issue** -- the conversation layer tasked the reasoner to send a slack message. Either the conversation layer needs to give better instructions, or both layers need the ability to talk on slack (via tools).

<img width="2799" height="1652" alt="image" src="https://github.com/user-attachments/assets/8ca50761-657e-472a-ac30-358a230c2cea" />

pavel-kirienko commented on src/bro/memory.py:
````diff
@@ -0,0 +1,119 @@
+```
+
+
+"""
````
I think it would be really beneficial to provide concrete examples how the memory can be used to store potentially very large amounts of data as files on the local filesystem, such that the memory engine here is only used to memorize file locations instead of their full contents. For example, "the Bishop financial forecast is stored in file such and such", "bank transaction exports are here and there", etc.

Also please trim blanks.

pavel-kirienko commented on src/bro/memory.py:
```diff
@@ -0,0 +1,119 @@
+},
+{
+    "text": "The company address is at Tallinn, Estonia", 
+    "tags": ["episodic", "semantics"],
```
The example given appears to belong to semantic memory, not episodic. "I visited the company in Tallinn, Estonia" would be a better fit, but then there may be better examples not involving Tallinn and companies. Like "Yesterday I ate ice cream at 1 am while petting a lion".

pavel-kirienko commented on src/bro/memory.py:
```diff
@@ -0,0 +1,119 @@
+},
+{
+    "text": "What is the user preferred language for communication?",
+    "sectors": ["episodic", "reflective", "semantic"],
```
The first and the last examples seem to be a poor fit for their sectors.

songmeo replied:
when I added the first example, I consider that the information is provided in for example wiki documents. Those documents are usually stored in procedural sectors. Reflective sector is because the memory has auto-reflection mode so I thought after the model reflecting on the wiki documents, it might add that tag to the info.

pavel-kirienko commented on src/bro/memory.py:
```diff
@@ -0,0 +1,119 @@
+    tier="smart",
+    embeddings={"provider": "openai", "apiKey": os.getenv("OPENAI_API_KEY"), "model": "text-embedding-3-small"},
+    reflection={"enabled": True, "intervalMinutes": 1440, "minMemories": 10},  # Daily
+)
```
Remove the singleton and use a class instead, similar to the other modules here. Singletons are bad for composability, testing, and coupling.

pavel-kirienko commented on src/bro/memory.py:
Entities should be designed from the client's perspective, not from the perspective of implementation. This module implements memory, so it should be named accordingly (I think just `memory.py` would suffice). The fact that it uses OpenMemory is an unimportant implementation detail that should be hidden from the client.

pavel-kirienko commented on src/bro/memory.py:
```diff
@@ -0,0 +1,119 @@
+)
+
+
+def memory_handler(name: str, args: str) -> str:
```
Split into distinct methods in the class

pavel-kirienko commented on src/bro/memory.py:
```diff
@@ -0,0 +1,119 @@
+def memory_handler(name: str, args: str) -> str:
+    match name, args:
+        case ("remember", {"text": text, "tags": tags}):
+            _logger.info("Adding memory...")
```
To make logging entries useful, always add relevant context into them, but not to the point where it becomes unfit for human comprehension. Here, you definitely need to list the tags, and provide at least the first 200 symbols from the text.

pavel-kirienko commented on src/bro/memory.py:
```diff
@@ -0,0 +1,119 @@
+
+from openmemory import OpenMemory
+
+from bro.brofiles import MEMORY_FILE
```
Do not import directly, pass via ctor. Minimize coupling.

pavel-kirienko commented on src/bro/brofiles.py:
```diff
@@ -12,3 +12,4 @@
 SNAPSHOT_FILE = BRODIR_LOCAL / "state.bro.json"
 LOG_FILE = BRODIR_LOCAL / "bro.log"
 DB_FILE = BRODIR_LOCAL / "bro.db"
+MEMORY_FILE = BRODIR_LOCAL / "memory.db"
```
Since we have more than one db file at this point, I'm thinking the previous `DB_FILE` needs a more descriptive name. Currently we use it only for logs, so we can call it log DB?

songmeo committed:
rename to memory only

songmeo committed:
rename LOG_DB

songmeo committed:
rename slack env vars to use BRO_SLACK_* prefix

- SLACK_BOT_TOKEN -> BRO_SLACK_BOT_TOKEN
- SLACK_APP_TOKEN -> BRO_SLACK_APP_TOKEN
- BRO_USER_ID -> BRO_SLACK_USER_ID

songmeo committed:
move memory to a class

songmeo committed:
decouple memory handler

songmeo committed:
update README, token generation

songmeo committed:
add file location storage instruction to memory

songmeo committed:
fix minor

songmeo committed:
remove unneccessary log
